scanpy/get/get.py

```diff
@@ -437,3 +437,141 @@ def _set_obs_rep(adata, val, *, use_raw=False, layer=None, obsm=None, obsp=None)
+
+
+def split_by(
+    adata: AnnData,
+    key: str,
+    groups: Optional[Union[GroupsList, GroupsLists, GroupsDict]] = None,
+    others_key: Optional[str] = None,
+    axis: int = 0,
+    copy: bool = False,
+) -> Dict[str, AnnData]:
+    """\
+    Split adata by obs key.
+
+    Params
+    ------
+    adata
+        AnnData object to split.
+    key
+        A key from `.obs` or to `.var` depending on `axis`
+        to use for splitting `adata`.
+    groups
+        Specifies which groups to select from adata `key`.
+        If `None`, `adata` will be split with all values from `key`.
+        It can be a list of groups' names, a list of lists of groups to aggregate,
+        a dict of lists of groups to aggregate.
+    others_key
+        If not `None`, the returned dict will have an additional key with
+        this name and the adata subset object with all groups not specified
+        in `groups` as a value.
+    axis
+        Axis of `adata` to split by `.obs` or `.var` `key`.
+    copy
+        If `True`, all split AnnData objects are copied; otherwise,
+        the returned dict will have views.
+
+    Returns
+    -------
+    A dictionay with AnnData objects.
+    If names of the split AnnData objects are not specified
+    (a list of lists in `groups`), they will be created by joining the groups' names.
+
+    Examples
+    --------
+    Split by all values in an `.obs` `key`:
+
+    >>> adata = sc.datasets.pbmc68k_reduced()
+    >>> adatas = sc.get.split_by(adata, 'bulk_labels')
+    >>> adatas
+    {'CD14+ Monocyte': View of AnnData object with n_obs × n_vars = 129 × 765
+     ...,
+     'CD19+ B': View of AnnData object with n_obs × n_vars = 95 × 765
+     ...,
+     ...
+    }
+
+    Select only specific groups from `.obs` `key`:
+
+    >>> adata = sc.datasets.pbmc68k_reduced()
+    >>> adatas = sc.get.split_by(adata, 'bulk_labels', ['CD14+ Monocyte', 'CD34+'])
+    >>> adatas
+    {'CD14+ Monocyte': View of AnnData object with n_obs × n_vars = 129 × 765
+     ...,
+     'CD34+': View of AnnData object with n_obs × n_vars = 13 × 765
+     ...
+    }
+
+    Aggreagte some groups from `.obs` `key`, put all others to `others_key`:
+
+    >>> adata = sc.datasets.pbmc68k_reduced()
+    >>> adatas = sc.get.split_by(adata, 'bulk_labels',
+                                 dict(some=['CD14+ Monocyte', 'CD34+']),
+                                 others_key='others')
+    >>> adatas
+    {'some': View of AnnData object with n_obs × n_vars = 142 × 765
+     ...,
+     'others': View of AnnData object with n_obs × n_vars = 558 × 765
+     ...
+    }
+    """
+    if key not in adata.obs:
+        raise ValueError(f"No {key} in .obs.")
+
+    if axis not in (0, 1):
+        raise ValueError("axis should be 0 or 1 only.")
+
+    attr_by_key = adata.obs[key] if axis == 0 else adata.var[key]
+    select = [slice(None), slice(None)]
+
+    adatas = {}
+
+    if groups is None:
+        groups = np.unique(attr_by_key)
```

Comment:
This should be faster for categoricals (since pandas knows to just look at the categories)

```suggestion
        groups = attr_by_key.unique()
```

Reply:
Changed as you suggested.